stdlib/zipfile/__init__.pyi

```diff
@@ -106,24 +120,50 @@ class ZipFile:
     compresslevel: int | None  # undocumented
     mode: _ZipFileMode  # undocumented
     pwd: bytes | None  # undocumented
+    # metadata_encoding is new in 3.11
     if sys.version_info >= (3, 11):
         @overload
         def __init__(
             self,
             file: StrPath | IO[bytes],
+            mode: _ZipFileMode = "r",
+            compression: int = 0,
+            allowZip64: bool = True,
+            compresslevel: int | None = None,
+            *,
+            strict_timestamps: bool = True,
+            metadata_encoding: str | None = None,
+        ) -> None: ...
+        # metadata_encoding is only allowed for read mode
+        @overload
+        def __init__(
+            self,
+            file: StrPath | _ZipReadable,
             mode: Literal["r"] = "r",
             compression: int = 0,
             allowZip64: bool = True,
             compresslevel: int | None = None,
             *,
             strict_timestamps: bool = True,
-            metadata_encoding: str | None,
+            metadata_encoding: str | None = None,
         ) -> None: ...
         @overload
         def __init__(
             self,
-            file: StrPath | IO[bytes],
-            mode: _ZipFileMode = "r",
+            file: StrPath | _ZipTellable | _ZipWritable,
```

Comment:
While `_ZipTellable` won't immediately error, I don't think the resulting object can be used for anything right? Unless I'm missing some use case, I think we should keep it `StrPath | _ZipWritable`

Reply:
Oh in fact it's definitely wrong, because `__del__` will trigger `close` will trigger a write attempt

Reply:
Let me fix this.

Reply:
https://github.com/python/typeshed/pull/12119/commits/d98196ab5128bc54e5442dff57f1620eb8a67272

Reply:
@hauntsaninja Should be resolved now.